app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb

```diff
@@ -7,69 +7,154 @@ def initialize(
+          @identity_name_from_token = decoded_token[token_id_field_secret]
+          if @identity_name_from_token.blank?
+            raise Errors::Authentication::AuthnJwt::NoSuchFieldInToken, token_id_field_secret
+          end
 
-        def identity_field_variable
-          @identity_field_variable ||= @resource_class[token_id_field_variable_id]
+          @logger.debug(
+            LogMessages::Authentication::AuthnJwt::FoundJwtFieldInToken.new(
+              token_id_field_secret,
+              @identity_name_from_token
+            )
+          )
+          @identity_name_from_token
         end
 
         def token_id_field_secret
           return @token_id_field_secret if @token_id_field_secret
 
-          @token_id_field_secret = conjur_secret(token_id_field_variable_id)
-          @logger.info(LogMessages::Authentication::AuthnJwt::RetrievedResourceValue.new(@token_id_field_secret, token_id_field_variable_id))
+          @token_id_field_secret = @fetch_required_secrets.call(
+            resource_ids: [token_id_field_variable_id]
+          )[token_id_field_variable_id]
+
+          @logger.info(
+            LogMessages::Authentication::AuthnJwt::RetrievedResourceValue.new(
+              @token_id_field_secret,
+              token_id_field_variable_id
+            )
+          )
           @token_id_field_secret
         end
 
         def token_id_field_variable_id
           @token_id_field_variable_id ||= "#{@authentication_parameters.authn_jwt_variable_id_prefix}/#{TOKEN_APP_PROPERTY_VARIABLE}"
         end
 
-        def conjur_secret(secret_id)
-          @fetch_required_secrets.call(resource_ids: [secret_id])[secret_id]
+        def decoded_token
+          @authentication_parameters.decoded_token
+        end
+
+        def token_id_field_variable
+          @token_id_field_variable ||= @resource_class[token_id_field_variable_id]
+        end
+
+        def fetch_identity_path
+          identity_path
+        end
+
+        def identity_path
+          @identity_path ||= @fetch_identity_path.call(authentication_parameters: @authentication_parameters)
+        end
+
+        def identity_name_from_token
+          @identity_name_from_token || fetch_identity_name_from_token
+        end
+
+        def add_prefix_path_to_identity_name
+          return @identity_name_with_prefix_path if @identity_name_with_prefix_path
+
+          if identity_path.blank?
+            @identity_name_with_prefix_path ||= identity_name_from_token
+          else
+            @identity_name_with_prefix_path ||=
+              @add_prefix_to_identity.call(
+                identity_prefix: identity_path,
+                identity: identity_name_from_token
+              )
+          end
+
+          @identity_name_with_prefix_path
+        end
+
+        def identity_name_with_prefix_path
+          @identity_name_with_prefix_path ||= add_prefix_path_to_identity_name
+        end
+
```

Comment:
Here is another example that should be simplified and can remove the `add_prefix_path_to_identity_name` method altogether. You may be aware of others following this same pattern that I haven't spotted. These should be cleaned up.

```suggestion
           def identity_name_with_prefix_path
          @identity_name_with_prefix_path ||=
            if identity_path.blank?
              identity_name_from_token
            else
              @add_prefix_to_identity.call(
                identity_prefix: identity_path,
                identity: identity_name_from_token
              )
            end
        end

```

Reply:
Nice , thanks :)